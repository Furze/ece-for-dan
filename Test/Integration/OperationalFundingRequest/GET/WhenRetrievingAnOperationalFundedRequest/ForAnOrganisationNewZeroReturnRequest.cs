using System;
using System.Collections.Generic;
using System.Linq;
using Bard;
using Events.Integration.Protobuf.Entitlement;
using MoE.ECE.Domain.Event.OperationalFunding;
using MoE.ECE.Domain.Read.Model.OperationalFunding;
using Moe.ECE.Events.Integration;
using MoE.ECE.Integration.Tests.Chapter;
using MoE.ECE.Integration.Tests.Infrastructure;
using Shouldly;
using Xunit;
using Xunit.Abstractions;

namespace MoE.ECE.Integration.Tests.OperationalFundingRequest.GET.WhenRetrievingAnOperationalFundedRequest
{
    public class ForAnOrganisationNewZeroReturnRequest : IntegrationTestBase<ECEStoryBook, ECEStoryData>
    {
        public ForAnOrganisationNewZeroReturnRequest(RunOnceBeforeAllTests testSetUp, ITestOutputHelper output,
            TestState<ECEStoryBook, ECEStoryData> testState) : base(testSetUp, output, testState)
        {
        }

        private Guid _businessEntityId;

        protected override void Arrange() =>
            Given
                .A_rs7_has_been_created()
                .An_rs7_is_ready_for_internal_ministry_review()
                .GetResult(data => _businessEntityId = data.Rs7Model.BusinessEntityId.GetValueOrDefault());

        protected override void Act() =>
            When.Get($"api/operational-funding-requests?business-entity-id={_businessEntityId}&revision-number=1");

        [Fact]
        public void ThenADomainEventShouldBePublished() =>
            A_domain_event_should_be_fired<OperationalFundingRequestCreated>();

        [Fact]
        public void ThenAnIntegrationEventShouldBeCalculated()
        {
            var integrationEvent = An_integration_event_should_be_fired<EntitlementCalculated>();

            integrationEvent
                .BusinessEntityType
                .ShouldBe(Constants.BusinessEntityTypes.Rs7);
        }

        [Fact]
        public void ThenTheResponseShouldBeOk() => Then.Response.ShouldBe.Ok();

        [Fact]
        public void ThenTheResponseShouldContain1Item()
        {
            var result = Then.Response.Content<ICollection<OperationalFundingRequestModel>>();

            result.Count.ShouldBe(1);
        }

        [Fact]
        public void ThenTheResponseShouldContainMatchingAdvanceMonths()
        {
            var result = Then.Response.Content<ICollection<OperationalFundingRequestModel>>();

            result.ShouldNotBeNull();
            result.ElementAt(0).MatchingAdvanceMonths.ShouldNotBeNull();
            result.ElementAt(0).MatchingAdvanceMonths?.Count.ShouldBe(4);
            result.ElementAt(0).MatchingAdvanceMonths?.ElementAt(0).MonthNumber.ShouldBe(10);
        }

        [Fact]
        public void ThenTheResponseShouldContainNoFundingComponents()
        {
            var result = Then.Response.Content<ICollection<OperationalFundingRequestModel>>();

            result.ShouldNotBeNull();
            result.ElementAt(0).EntitlementMonths.ShouldNotBeNull();
            result.ElementAt(0).EntitlementMonths?.ElementAt(0).FundingComponents.ShouldNotBeNull();
            result.ElementAt(0).EntitlementMonths?.ElementAt(0).FundingComponents?.Count.ShouldBe(0);
        }
    }
}