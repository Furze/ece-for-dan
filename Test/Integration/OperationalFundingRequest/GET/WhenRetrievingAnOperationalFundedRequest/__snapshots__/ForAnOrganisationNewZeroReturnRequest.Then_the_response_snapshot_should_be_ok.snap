[
  {
    "EntitlementMonths": [
      {
        "MonthName": "February 2020",
        "MonthNumber": 2,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 0,
        "SessionalNonCertificatedTeacherHours": 0,
        "TotalWorkingDays": 0,
        "WashUpPlusTen": -2800.8,
        "WashUpUnderTwo": -9343.35,
        "WashUpTwoAndOver": -4959.75,
        "WashUpTwentyHours": -7750.05,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 24853.95,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": -24853.95,
        "FundingComponents": []
      },
      {
        "MonthName": "March 2020",
        "MonthNumber": 3,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 0,
        "SessionalNonCertificatedTeacherHours": 0,
        "TotalWorkingDays": 0,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "April 2020",
        "MonthNumber": 4,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 0,
        "SessionalNonCertificatedTeacherHours": 0,
        "TotalWorkingDays": 0,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "May 2020",
        "MonthNumber": 5,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 0,
        "SessionalNonCertificatedTeacherHours": 0,
        "TotalWorkingDays": 0,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": []
      }
    ],
    "OrganisationId": 114895,
    "BusinessEntityId": "ca2b66f7-4676-457e-8bf7-98e40540df05",
    "RequestId": "102005",
    "FundingPeriodMonth": "July",
    "FundingYear": 2021,
    "RevisionNumber": 1,
    "TotalWashUp": -24853.95,
    "TotalAdvance": 0.0,
    "AdvanceMonths": [
      {
        "MonthName": "July 2020",
        "MonthNumber": 7,
        "Year": 2020,
        "ParentLedWorkingDays": 0,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": null,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "August 2020",
        "MonthNumber": 8,
        "Year": 2020,
        "ParentLedWorkingDays": 0,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": null,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "September 2020",
        "MonthNumber": 9,
        "Year": 2020,
        "ParentLedWorkingDays": 0,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": null,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "October 2020",
        "MonthNumber": 10,
        "Year": 2020,
        "ParentLedWorkingDays": 0,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": null,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 0,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      }
    ],
    "MatchingAdvanceMonths": [
      {
        "MonthName": "February",
        "MonthNumber": 2,
        "Year": 2020,
        "ParentLedWorkingDays": null,
        "AllDayWorkingDays": 20,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 2800.8,
        "AmountPayableUnderTwo": 9343.35,
        "AmountPayableTwoAndOver": 4959.75,
        "AmountPayableTwentyHours": 7750.05,
        "TotalDays": null,
        "TotalAdvance": null,
        "FundingComponents": [
          {
            "SessionType": "AllDay",
            "FundingComponentType": "PlusTen",
            "StartDate": "0001-01-01T00:00:00",
            "Amount": 2800.8,
            "Rate": 10.1,
            "RateName": "",
            "FundedChildHours": 30,
            "OperatingDays": 20
          },
          {
            "SessionType": "AllDay",
            "FundingComponentType": "UnderTwo",
            "StartDate": "0001-01-01T00:00:00",
            "Amount": 9343.35,
            "Rate": 10.1,
            "RateName": "",
            "FundedChildHours": 30,
            "OperatingDays": 20
          },
          {
            "SessionType": "AllDay",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "0001-01-01T00:00:00",
            "Amount": 4959.75,
            "Rate": 10.1,
            "RateName": "",
            "FundedChildHours": 30,
            "OperatingDays": 20
          },
          {
            "SessionType": "AllDay",
            "FundingComponentType": "TwentyHours",
            "StartDate": "0001-01-01T00:00:00",
            "Amount": 7750.05,
            "Rate": 10.1,
            "RateName": "",
            "FundedChildHours": 30,
            "OperatingDays": 20
          }
        ]
      }
    ],
    "Equity": null
  }
]
